openmc/search.py

```diff
@@ -0,0 +1,197 @@
+from collections import Callable
+from numbers import Real
+
+import openmc
+import openmc.model
+import openmc.checkvalue as cv
+
+
+_SCALAR_BRACKETED_METHODS = ['brentq', 'brenth', 'ridder', 'bisect']
+
+
+def _search_keff(guess, target, model_builder, model_args, print_iterations,
+                 print_output, guesses, results):
+    """Function which will actually create our model, run the calculation, and
+    obtain the result. This function will be passed to the root finding
+    algorithm
+
+    Parameters
+    ----------
+    guess : Real
+        Current guess for the parameter to be searched in `model_builder`.
+    target_keff : Real
+        Value to search for
+    model_builder : collections.Callable
+        Callable function which builds a model according to a passed
+        parameter. This function must return an openmc.model.Model object.
+    model_args : dict
+        Keyword-based arguments to pass to the `model_builder` method.
+    print_iterations : bool
+        Whether or not to print the guess and the resultant keff during the
+        iteration process.
+    print_output : bool
+        Whether or not to print the OpenMC output during the iterations.
+    guesses : Iterable of Real
+        Running list of guesses thus far, to be updated during the execution of
+        this function.
+    results : Iterable of Real
+        Running list of results thus far, to be updated during the execution of
+        this function.
+
+    Returns
+    -------
+    float
+        Value of the model for the current guess compared to the target value.
+
+    """
+
+    # Build the model
+    model = model_builder(guess, **model_args)
+
+    # Run the model and obtain keff
+    keff = model.run(output=print_output)
+
+    # Close the model to ensure HDF5 will allow access during the next
+    # OpenMC execution
+    model.close()
+
+    # Record the history
+    guesses.append(guess)
+    results.append(keff)
+
+    if print_iterations:
+        text = 'Iteration: {}; Guess of {:.2e} produced a keff of ' + \
+            '{:1.5f} +/- {:1.5f}'
+        print(text.format(len(guesses), guess, keff[0], keff[1]))
+
+    return (keff[0] - target)
+
+
+def search_for_keff(model_builder, initial_guess=None, target=1.0,
+                    bracket=None, model_args={}, tol=None,
```

Comment:
Having a mutable default value for an argument (in this case `model_args`) is a [no-no](https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html#using-a-mutable-default-value-as-an-argument).